MBBSEmu/Memory/MemoryAllocator.cs

```diff
@@ -0,0 +1,215 @@
+using NLog;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System;
+
+namespace MBBSEmu.Memory
+{
+  /// <summary>
+  ///   A Memory Allocator that implements malloc/free.
+  /// </summary>
+  public class MemoryAllocator
+  {
+    /// <summary>
+    ///   Represents a block of memory.
+    /// </summary>
+    private class MemoryBlock
+    {
+      /// <summary>
+      ///   Offset within the segment where this memory begins.
+      /// </summary>
+      public ushort Offset { get; set; }
+      /// <summary>
+      ///   Size in bytes of this memory block.
+      /// </summary>
+      public uint Size { get; set; }
+
+      /// <summary>
+      ///   Offset within the segment where this memory ends.
+      /// </summary>
+      public uint EndOffset { get => Offset + Size; }
+    }
+
+    /// <summary>
+    ///   List of free memory blocks we currently have.
+    /// </summary>
+    private LinkedList<MemoryBlock> _freeBlocks = new();
+    /// <summary>
+    ///   Stores all of our allocated memory blocks.
+    ///
+    ///   Key is the pointer, Value is the size of the allocated memory block.
+    /// </summary>
+    private ConcurrentDictionary<FarPtr, uint> _allocatedBlocks = new();
+
+    /// <summary>
+    ///   Lowest memory address that this allocator can return as a valid pointer.
+    /// </summary>
+    public FarPtr BasePointer { get; init; }
+
+    /// <summary>
+    ///   Total capacity in bytes that this memory allocator can allocate.
+    ///
+    ///   Could allocate less than this value depending on the fragmentation level.
+    /// </summary>
+    public uint Capacity { get; init; }
+
+    /// <summary>
+    ///   How many remaining bytes are available to be allocated.
+    /// </summary>
+    public uint RemainingBytes { get; set; }
+
+    public ILogger Logger { get; init; }
+
+    /// <summary>
+    ///   The number of free blocks being tracked.
+    ///
+    ///   Only used for testing.
+    /// </summary>
+    public int FreeBlocks { get => _freeBlocks.Count; }
+
+    /// <summary>
+    ///   Constructs a new MemoryAllocator.
+    ///
+    ///   basePointer + capacity must be smaller than a segment size, to ensure it will not overflow
+    ///   allocations into the next segment.
+    /// </summary>
+    /// <param name="logger">logger</param>
+    /// <param name="basePointer">lowest memory address returnable from malloc</param>
+    /// <param name="capacity">size in bytes that this allocator can draw memory from</param>
+    public MemoryAllocator(ILogger logger, FarPtr basePointer, uint capacity)
+    {
+      if ((basePointer.Offset & 0x1) != 0)
+        throw new ArgumentException("basePointer must be aligned on a word boundary");
+
+      var endOffset = basePointer.Offset + capacity;
+      if (endOffset > 0x10000)
```

Comment:
This is correct. It computes off + capacity to see if it overflows, and if you start from 0 and have a capacity of 65536 (whole segment), you'll get 0x10000 which is a valid value. Anything > this would be an overflow